### PR TITLE
Fixed freezing while playing music from file objects

### PR DESCRIPTION
**Pull request by dlon**
#548

The problem is primarily `_pg_rw_read_th`. Calling fileobj.read() (and possibly Py_DECREF) makes Python switch to the other thread and often causes the other thread to wait forever to acquire the mixer mutex.

My solution was to store the file descriptor and use this instead of the Python method, if it's available. Are there any drawbacks to this?

### CODE REVIEW

File: src_c/rwobject.c [hunk 1]
```diff
@@ -398,6 +402,17 @@ _pg_rw_seek(SDL_RWops *context, Sint64 offset, int whence)
     Sint64 retval;
 #endif
 
+    if (helper->fileno != -1) {
+        if (!(offset == 0 &&
```
Comment (illume) [inline src_c/rwobject.c]: Is the logic here correct?

It seems lseek could be called twice here in the case where offset != 0 && whence != SEEK_CUR, and lseek doesn't return -1?

**Comment by illume** [comment]
Thanks for this.

- I wonder if we have a unit test for this already? 
- I guess this leave the bug there for file objects that don't have file descriptors.

File: src_c/rwobject.c [hunk 2]
```diff
@@ -435,6 +450,15 @@ _pg_rw_read(SDL_RWops *context, void *ptr, size_t size, size_t maxnum)
     size_t retval;
 #endif /* IS_SDLv2 */
 
+    if (helper->fileno != -1) {
+        retval = read(helper->fileno, ptr, size * maxnum);
+        if (retval == -1) {
+            return -1;
+        }
+        retval /= size;
```
Comment (illume) [inline src_c/rwobject.c]: I guess it's possible that read could return less than a multiple of size. It's possible read could return 1 byte for example 4 times in a row. I'm not sure how to handle this case really, and it's also quite unlikely.

File: src_c/rwobject.c (hunk 1 above)
Comment (dlon) [reply src_c/rwobject.c]: True. It did do the right thing but that extra lseek was unnecessary.

File: src_c/rwobject.c (hunk 2 above)
Comment (dlon) [reply src_c/rwobject.c]: Seems to be true of the other call as well, so it shouldn't be any worse at least.

**Pull request merged**